source/common/access_log/access_log_formatter.h

```diff
@@ -95,14 +95,33 @@ class FormatterImpl : public Formatter {
                      const StreamInfo::StreamInfo& stream_info) const override;
 
 private:
-  std::vector<FormatterPtr> formatters_;
+  std::vector<FormatterProviderPtr> providers_;
+};
+
+class JsonFormatterImpl : public Formatter {
+public:
+  JsonFormatterImpl(const std::map<const std::string, const std::string>& format_mapping);
+
+  // Formatter::format
+  std::string format(const Http::HeaderMap& request_headers,
+                     const Http::HeaderMap& response_headers,
+                     const Http::HeaderMap& response_trailers,
+                     const StreamInfo::StreamInfo& stream_info) const override;
+
+  std::unordered_map<std::string, std::string>
+  toMap(const Http::HeaderMap& request_headers, const Http::HeaderMap& response_headers,
```

Comment:
If you want this accessible in tests, I suggest creating a test peer. I'm not sure why we need to test this internal behavior though, I'd skip it; it's pretty trivial.

Reply:
I took another look at this. If we stop using this function to test and make it private, then we'll be relying on either
1. parsing the JSON string in order to do our asserts or
2. hardcoding an expected JSON string and comparing the raw string

I'll attempt `1`, and see how it goes. I suppose it's a more accurate test anyway. I was worried `2` might be susceptible to the order of keys changing and therefore randomly failing tests

Reply:
yep, 1 sounds good!

Reply:
Done